Added poll insertion pass

anmolsahoo25 opened the pull request:
This patch adds a Mach op called *Ipoll* along with a Mach-to-Mach
pass for inserting polls according to the balanced polling
algorithm

anmolsahoo25 committed:
Added poll insertion pass

This patch adds a Mach op called *Ipoll* along with a Mach-to-Mach
pass for inserting polls according to the balanced polling
algorithm

kayceesrk commented on runtime/amd64.S:
```diff
@@ -448,6 +448,110 @@ CFI_STARTPROC
 ENDFUNCTION(G(caml_allocN))
 
+/* Polling */
+FUNCTION(G(caml_call_poll))
```
Rather than copy-pasting this large piece of code, use a macro definition to take in the function to call as an argument and shared the code between `caml_call_gc` and `call_call_poll`.

stedolan replied:
I might have missed something, but why is this function needed at all? Why not just call caml_call_gc here?

kayceesrk commented on asmcomp/amd64/emit.mlp:
```diff
@@ -309,7 +310,11 @@ let emit_call_gc gc =
     spacetime_before_uninstrumented_call ~node_ptr ~index
   end;
-  emit_call "caml_call_gc";
+  if (not gc.is_poll) then begin
```
Use `if (gc.is_poll)` rather than checking for negative condition.

kayceesrk commented on runtime/signals_nat.c:
```diff
@@ -98,6 +98,37 @@ void caml_garbage_collection(void)
 
+void caml_poll(void)
+{
+  frame_descr* d;
```
Why does `caml_poll` need to examine frame descriptor at all? Should `caml_poll` directly call one of the functions that processes pending signals? For example, having a quick scan through the code, I would think that `caml_poll` should directly call `caml_raise_if_exception(caml_do_pending_actions_exn ())` as in https://github.com/ocaml/ocaml/blob/trunk/runtime/minor_gc.c#L499.

But that raises an interesting problem. Because poll points are inserted by the compiler even in non-allocating code, exceptions may magically appear out of non-allocating code. Currently, such exceptions can only come from allocations, which can be guarded. My suggestion would be to introduce a global variable (registered as root) to store the exception and in the next allocation point (not poll point), raise that exception. [CC @stedolan]

anmolsahoo25 replied:
We need the frame descriptor for the spacetime/memprof functions, because they require the pointers to the allocations etc. Just looking at the spacetime/memprof parts to understand this better.

kayceesrk replied:
The idea is not to call the allocation code at all. In that case, you don't need to examine the frame descriptors. See original comment:

> I would think that caml_poll should directly call caml_raise_if_exception(caml_do_pending_actions_exn ()) as in https://github.com/ocaml/ocaml/blob/trunk/runtime/minor_gc.c#L499.

kayceesrk replied:
Given the trickiness with asynchronous exceptions, the right thing to do would be to define a function that only does the part of `caml_do_pending_actions_exn` that is known not to raise exceptions https://github.com/ocaml/ocaml/blob/trunk/runtime/signals.c#L293-L302. This should fix the out-of-memory bugs upstream. For multicore, we can add the poll for interrupts in this new function.

anmolsahoo25 replied:
1. Even I was thinking the poll function can be changed to only perform operations which do not raise exceptions.
2. Given that the poll can trigger a minor collection, I think spacetime/memprof might need to be signalled, regarding the collections? Not very familiar with how memprof/spacetime works, so need to look into this.

kayceesrk replied:
Reg 2, IINM both spacetime and memprof look at allocations and not collections. If a collection was necessary, then the collection code itself should handle the necessary logic for spacetime and memprof.

stedolan replied:
Regarding exceptions appearing out of nowhere: I don't think exceptions magically appearing in non-allocating code is really worse than exceptions magically appearing in allocations, as happens today. (Both are awful, though, but that's a separate issue!)

anmolsahoo25 committed:
Fixes memprof

This patch adds the memprof function call in the poll function
which fixes the memprof test failing.

anmolsahoo25 commented:
I have been able to narrow the problem to poll being called inside the memprof callback, which is causing some value to be corrupted.

stedolan commented on asmcomp/emitaux.ml:
```diff
@@ -198,9 +198,9 @@ let emit_frames a =
                 alloc_words - 1 <= Config.max_young_wosize &&
-                Config.max_young_wosize <= 256);
+                Config.max_young_wosize <= 256));
         a.efa_8 (alloc_words - 2)) dbg;
```
The assertion that alloc_words is between 2 and 257 is to make sure that `efa_8 (alloc_words - 2)` doesn't overflow. This changes the assertion without changing the format, so this isn't going to work.

Why do polling points need a Dbg_alloc frame anyway?

stedolan commented:
There was an issue with the original version of Dhruv's Cmm-level patch, because there are some positions where it's invalid to insert polling points. I think this patch might have the same bug.

One example of this is array accesses. The compiled form of a.(i) * 7 consists of:

  1. A bounds check that `0 <= i < Array.length a`
  2. An address computation of `a + i * size_addr`
  3. A load of the computed address
  4. A multiplication by 7

The issue is that there can be no polling point between 2 and 3. The reason is that after 2, there's an address in a register that points somewhere inside the array. The garbage collector doesn't know how to deal with such addresses, and it can't update the address correctly if it moves the array. So, we need to avoid ever calling the garbage collector when such internal pointers are lying around.

I think the current version of this patch might put a polling point between 2 and 3, which will cause either a compile-time assertion failure (when it tries to generate a bad frametable), or worse, a weird hard-to-debug GC bug.

anmolsahoo25 commented:
As you mentioned with regards to inserting polls where the address computation is ongoing, I have taken care of that. Since the poll insertion happens after the register allocation pass, I do have the information at which points addresses are live in the registers and I am not inserting polls there.

stedolan commented:
Thanks for the explanation! I see the is_addr_live check now, that makes sense.
